web/src/components/areagraph.js

```diff
@@ -1,466 +1,278 @@
-                .style('display', 'block');
-        }
-        if (this._selectedLayerIndex != null && this._selectedIndex != null && this._data[this._selectedIndex]) {
-            var selectedData = this.stack[this._selectedLayerIndex][this._selectedIndex];
-            this.markerElement
-                .attr('cx', this.x(this._data[this._selectedIndex].datetime))
-                .attr('cy', this.y(selectedData[1]))
-                .style('display', 'block')
-                .style('fill', function() {
-                    return fillColor.call(that, {
-                        key: that.stackKeys[that._selectedLayerIndex]
-                    })
-                })
-        } else {
-            this.markerElement.style('display', 'none')
-        }
+    if (mouseMoveHandler) {
+      mouseMoveHandler(layer[i].data._countryData, i);
     }
-    return this;
-}
-AreaGraph.prototype.displayByEmissions = function(arg) {
-    if (arg === undefined) return this._displayByEmissions;
-    else {
-        this._displayByEmissions = arg;
-        // Quick hack to re-render
-        // TODO: In principle we shouldn't be calling `.data()`
-        this
-            .data(this._originalData)
-            .render();
+  };
+  const handleLayerMouseOut = () => {
+    mouseOutTimeout = setTimeout(() => {
+      setSelectedLayerIndex(undefined);
+      if (mouseOutHandler) {
+        mouseOutHandler();
+      }
+      if (layerMouseOutHandler) {
+        layerMouseOutHandler();
+      }
+    }, 50);
+  };
+
+  return (
+    <g>
+      {stackedData.map((layer, ind) => (
+        <path
+          key={stackKeys[ind]}
+          className={`area layer ${stackKeys[ind]}`}
+          fill={fillColor(stackKeys[ind], displayByEmissions)}
+          style={{ cursor: 'pointer' }}
+          d={area(layer)}
+          onFocus={ev => handleLayerMouseMove(ev, layer, ind)}
+          onMouseOver={ev => handleLayerMouseMove(ev, layer, ind)}
+          onMouseMove={ev => handleLayerMouseMove(ev, layer, ind)}
+          onMouseOut={handleLayerMouseOut}
+          onBlur={handleLayerMouseOut}
+        />
+      ))}
+    </g>
+  );
+});
+
+const getCurrentTime = state =>
+  state.application.customDate || (state.data.grid || {}).datetime;
+
+// Regular production mode or exchange fill as a fallback
+const fillColor = (key, displayByEmissions) =>
+  modeColor[key] || (displayByEmissions ? 'darkgray' : `url(#areagraph-exchange-${key})`);
+
+const mapStateToProps = (state, props) => ({
+  colorBlindModeEnabled: state.application.colorBlindModeEnabled,
+  currentTime: getCurrentTime(state),
+  data: props.dataSelector(state),
+  displayByEmissions: state.application.tableDisplayEmissions,
+  electricityMixMode: state.application.electricityMixMode,
+  selectedIndex: state.application.selectedZoneTimeIndex,
+});
+
+const AreaGraph = ({
+  colorBlindModeEnabled,
+  currentTime,
+  data,
+  displayByEmissions,
+  electricityMixMode,
```

Comment:
`electricityMixMode` and `displayByEmissions` are two variables that in principle should be taken out of `AreaGraph` if we want the graph to be re-usable.
I was thinking, would it make sense to try to move those two out of the `AreaGraph`, and the replace the `data` prop with something less general that is agnostic of "exchanges", "countries" and "carbonintensity"?
Basically we would move `getGraphState` outside and pass directly `area` etc.. instead of `data`?

Let me know what you think. I think it could be worth it to clean up the code even more. Maybe we could do it in another PR?

Reply:
Yeah, that's a good point!

I was looking into it and it looks like a couple of hours work to make the component fully generic, so I thought we might actually want to defer it to a later time - I opened #2163 to capture the problem and suggest a solution.

From a technical point it's definitely something I'd like to get around to, but maybe it shouldn't block the rest of the React transition as its motivation is somewhat orthogonal to this effort.

If you'd like to see it addressed soon, I'd be happy to play with #2163 in one of the next few days.

WDYT?

Reply:
I think it'd be worth addressing soon while it's fresh in your mind.

Reply:
This comment has been addressed by https://github.com/tmrowco/electricitymap-contrib/pull/2171.